running tests in a container

tophj-ibm opened the pull request:
This provides a dockerfile for building an image to run the containerd tests.

Signed-off-by: Christopher Jones <tophj@linux.vnet.ibm.com>

few WIP issues:

- [X] optimize getting protobuf, which currently takes ~25minutes to build from source, which is done because google only provides amd64 downloads
- [X] `make integration` fails at the end because it tries to unmount the test dir which is a mounted vol
- [X] `make root-test` fails ~~because it runs of out of loop devices, this is~~ because previous tests devices are erroring out and not being freed because of :
```
failed to make btrfs filesystem (out: "ERROR: '/dev/loop3' is too small to make a usable filesystem\nERROR: minimum size for each btrfs device is 142606336\nbtrfs-progs v4.7.3\nSee http://btrfs.wiki.kernel.org for more information.\n\n")
```

stevvooe commented on Dockerfile:
```diff
@@ -0,0 +1,57 @@
+	&& make binaries install
+
+VOLUME /var/lib/containerd-test
+VOLUME /tmp
```
Is this necessary? I think we mostly expect a tmpfs here. Were there problems with that?

tophj-ibm replied:
It's the same issue that /var/lib/containerd-test was having, the overlay tests attempt to mount to /tmp but fail to.

crosbymichael replied:
Docker has a dind script that mounts cgroups and /tmp as tmpfs, I think we should run that as the entrypoint.

stevvooe commented on Dockerfile:
```diff
@@ -0,0 +1,57 @@
+VOLUME /var/lib/containerd-test
+VOLUME /tmp
+
+WORKDIR /go/src/github.com/containerd/containerd
```
Should the entry point be `make test` or `make` or `go test ./...`?

dnephin commented on Dockerfile:
```diff
@@ -0,0 +1,57 @@
+	&& cd /go/src/github.com/containerd \
+	&& git clone "https://github.com/containerd/containerd.git" \
+	&& cd containerd \
+	&& make binaries install
```
I generally avoid this "git checkout of the source repo" in a Dockerfile. I think it's safe to assume that this Dockerfile will only be used with a build context of a `containerd` checkout, so it should use `COPY . /go/src/github.com/containerd/containerd` or use a bind mount to get a source.

`make binaries install` can be done as part of the `ENTRYPOINT` or omitted entirely.

dnephin commented on Dockerfile:
```diff
@@ -0,0 +1,57 @@
+RUN set -x \
+	&& mkdir -p /go/src/github.com/opencontainers \
+	&& cd /go/src/github.com/opencontainers \
+	&& git clone "https://github.com/opencontainers/runc.git" \
```
I think this should use the pinned version from `RUNC.md` since that's what we document as being supported.

crosbymichael commented:
@tophj-ibm whats the status on this one now that the tests are passing?

dnephin commented:
I believe the 3 bullet points in the description are still relevant and need to be fixed. The tests aren't running using this image just yet, but we should be able to use it with #634

tophj-ibm commented:
@crosbymichael bullet points are still relevant, while the non-root tests pass, I'm still having issues with the root-ones.

I have a patch for downloading protobuf on amd64 only, but I haven't timed it yet and it doesn't look like it will be much faster. I have to change the tests to not unmount the directory probably or move it to a tmpfs for the `integration` test issue, and I believe the btrfs-progs package is bugged on the version of debian I am working with causing the `root-test` issue, I'm still investigating.

tophj-ibm commented:
Fixed the first two. Anyone have any ideas on the last point? No matter what I try or how big a mount I give it, `mkfs.btrfs` just complains about the size and errors out in a container.

crosbymichael commented:
Are you enabling btrfs by default?

tophj-ibm commented:
I'm not doing anything special. Building containerd with or without it doesn't make a difference. Does the host docker mount and graphdriver need to be btrfs?

crosbymichael commented:
It shouldn't.  I guess we need to figureout where in the code that error is coming from.

estesp commented:
@tophj-ibm the problem with item 3 is discussed here: https://github.com/moby/moby/issues/27886

If I add `-v /dev:/dev` to your `docker run` command the tests work. I found the following other changes make things smoother:

1. Add btrfs-tools and xfstools to your final section of the `Dockerfile` `RUN apt-get install`
2. Correct your `docker run` comment at the top of the `Dockerfile` to add `bash` as the command; otherwise you just get an error and exit instead of a shell. And of course add the `-v /dev:/dev` to handle the loop hotplug issue.
3. Not required, but adding the vim-tiny or some simple vi editor would make it easier to do simple activities inside the container shell :)

tophj-ibm commented:
Thanks for testing @estesp! I'll update those things

I found the underlying issue ended up being` ppc64le` has a higher threshold for a minimum btrfs size on mkfs, so the error message was accurate, and my attempts to test it with a larger volume were... incorrect. Upping the size of the file worked.

I think I'm close to being done with this, running into a small issue now with running out of space because the parallel tests all create filesystems that aren't being cleaned up immediately but I have a few things I can do.

dnephin commented on client_test.go:
```diff
@@ -132,9 +132,12 @@ func TestMain(m *testing.M) {
-			os.Exit(1)
+		// this directory is a volume in a container so don't remove
+		if len(os.Getenv("container")) == 0 {
+			if err := os.RemoveAll(defaultRoot); err != nil {
```
Maybe just ignore the remove error instead of skipping remove?

If someone uses a bind mount for this directory, or runs a shell and does multiple test runs in the same container,  they'll still want the files cleaned up.

tophj-ibm commented:
Got all the tests to pass, feel free to review!

tophj-ibm commented on snapshots/btrfs/btrfs_test.go:
```diff
@@ -29,20 +29,23 @@ func boltSnapshotter(t *testing.T) func(context.Context, string) (snapshots.Snap
 	return func(ctx context.Context, root string) (snapshots.Snapshotter, func() error, error) {
 
-		deviceName, cleanupDevice, err := testutil.NewLoopback(100 << 20) // 100 MB
+		deviceName, cleanupDevice, err := testutil.NewLoopback(650 << 20) // 650 MB
```
I can leave a code comment here if you want, but this was the minimum size I could get btrfs to work in a container on `ppc64le`. I tested with `s390x` and it worked with an even smaller value. I don't have the right hardware to test this with arm*

AkihiroSuda replied:
Why the required size significantly differs across architectures?
Is there ppc-specific bug in our code or in kernel?

tophj-ibm replied:
I couldn't find any information / bugs reported on this, but it's definitely a kernel issue and not a containerd one.

AkihiroSuda replied:
Could you open a github issue?

stevvooe replied:
Please fix this for systems that work under this size. This is way too big to be allocating on small hosts.

dnephin commented on script/setup/install-runc:
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+#
+# Builds and installs runc to /usr/local/go/bin based off
+# the commit defined in RUNC.md
```
There's a line in `.travis.yml` which installs runc, it should probably be replaced with `./script/setup/install-runc`

dnephin commented on script/setup/install-protobuf:
```diff
@@ -0,0 +1,27 @@
+#
+# Downloads and installs protobuf
+#
+set -eux -o pipefail
```
Remove `-x` by default? could be enabled with `$DEBUG`, but I'm not sure if it's necessary.

dnephin commented on script/setup/install-runc:
```diff
@@ -0,0 +1,15 @@
+# Builds and installs runc to /usr/local/go/bin based off
+# the commit defined in RUNC.md
+#
+set -eux -o pipefail
```
same about `-x`

dnephin commented on script/setup/install-protobuf:
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env bash
+#
+# Downloads and installs protobuf
```
There are a few lines in `.travis.yml` that could be replaced by `./script/setup/install-protobuf`

tophj-ibm commented:
added in the travis scripts too PTAL

AkihiroSuda commented on script/setup/install-runc:
```diff
@@ -0,0 +1,13 @@
+#
+set -eu -o pipefail
+
+RUNC_COMMIT=$(grep RUNC_COMMIT RUNC.md | cut -d "=" -f 2)
```
I'm -1 for forcibly parsing machine-unreadable document. Why not use TOML/YAML?

dnephin replied:
If we move the sha out from the docs when we have to duplicate it and remember to update both, or we can't document the supported version.

AkihiroSuda replied:
We can just put `Please refer to foo.toml` to `RUNC.md`

AkihiroSuda commented on client_test.go:
```diff
@@ -133,8 +133,11 @@ func TestMain(m *testing.M) {
 		if err := os.RemoveAll(defaultRoot); err != nil {
-			fmt.Fprintln(os.Stderr, "failed to remove test root dir", err)
-			os.Exit(1)
+			// this directory is a volume for container tests, so we don't care if it fails
```
Let's create the root within the volume and remove this code

dnephin replied:
I think we can just remove this code anyway, we don't need to change the path.  

There's no reason to `os.Exit(1)` if the RemoveAll fails, just ignore the error in all cases. We already do that in dozens of places in the test suite.

AkihiroSuda commented on Dockerfile:
```diff
@@ -0,0 +1,35 @@
+# This dockerfile is used to test containerd within a container
+#
+# usage:
+# 1.) docker build -t containerd-test -f Dockerfile .
```
s/Dockerfile/Dockerfile.test/ , as this Dockerfile is not suitable for other purposes

dnephin replied:
In a follow up we could add another stage which runs `script/setup/install-dev-tools`. That would make this Dockerfile usable for just about every project task. Then do you think we could keep it `Dockerfile` or `Dockerfile.dev` ?

AkihiroSuda commented:
Off-topic: I think we should register `https://hub.docker.com/u/containerd` before somebody else takes over, (even if we don't use Docker Hub).

tophj-ibm commented on script/setup/install-runc:
```diff
@@ -0,0 +1,13 @@
+#
+set -eu -o pipefail
+
+RUNC_COMMIT=$(grep opencontainers/runc vendor.conf | cut -d " " -f 2)
```
@AkihiroSuda what do you think about this?

AkihiroSuda replied:
Generally vendor.conf != RUNC.md, but I think it is ok to call it a day

AkihiroSuda commented on script/setup/install-runc:
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+#
+# Builds and installs runc to /usr/local/go/bin based off
+# the commit defined in RUNC.md
```
s/RUNC.md/vendor.conf

stevvooe commented:
What's the status of this PR? It seems like it creates a bunch of dependencies on docker, which seems pretty backwards for the goals of this project. I _really_ don't want to bring the testing experience of moby/moby to this project.

If we want to self-host test inside of a container for containerd, that is great, but we need to have a plan that doesn't create a dependency on docker.

estesp commented:
I don't think this was proposed to be a replacement for any of the development processes we are following today in containerd. We've said multiple times (in now closed issues) if someone wants to contribute a `Dockerfile` then that's fine, as long as it is maintained over time. If we think we need to move it to `contrib/` to make that clearer that's fine with me.

To me this is simply a convenience offering for those who want to leave their system unmodified and yet do development and test related activities with containerd in an isolated container. That was kind of one of the promises of that Docker thing. :)

tophj-ibm commented:
@stevvooe I see your concerns and I think @estesp had a good suggestion, what do you think about `contrib/`? It decreases visibility, but I think it addresses your issues regarding docker dependencies in this project.

I created this PR to make it easy to run containerd, and more specifically for me, all the tests, in a container. The dockerfile and subsequent scripts solves issues that I ran into while building and testing, including making sure specific directories such as `/var/lib/containerd-test` and `/tmp` were mounted, as well as providing an easy way to install the dependencies on other platforms where it isn't as simple as "download protobuf".

As for the state of this PR: all the tests pass, feel free to try it yourself. 

I am tracking down a probable bug with btrfs-progs where `ppc64le` needs a significantly larger minimum storage size (~650MB vs ~100MB) for mkfs.btrs, but I don't think that's a blocker here.

dnephin commented:
> it creates a bunch of dependencies on docker

There is no dependency on docker. The Dockerfile can be added as a convenience for contributors, but there is no requirement to use it. I will replace my adhoc Dockerfile with this one if it merges. I sure there are other contributors who will benefit.

> I really don't want to bring the testing experience of moby/moby to this project.

No one does.

There are thousands of projects that use Docker as a development tool, many of which have much better dev/testing experiences than the moby/moby project.

> self-host test inside of a container for containerd, that is great

This sounds like a lot more work to develop and maintain. We already have a tool for this.

stevvooe commented:
Fair enough. Place these in contrib/ and let’s keep them out of the Makefile.

This process needs to be updated in the building.md file so people know how to use it.

stevvooe commented on client_test.go:
```diff
@@ -134,7 +134,6 @@ func TestMain(m *testing.M) {
 		}
 		if err := os.RemoveAll(defaultRoot); err != nil {
 			fmt.Fprintln(os.Stderr, "failed to remove test root dir", err)
-			os.Exit(1)
```
This changes the behavior of testmain.

dnephin replied:
We had some discussion about this previously in this PR. I suggested that there shouldn't be any reason to exit(1) here. Am I mistaken? What's the reason for an `Exit(1)` if the remove fails?

stevvooe replied:
Ok, can we fix in another PR? This is an area where we previously exited but do not now. I have no clue why this is the way it is, so let's see if we can figure that out.

tophj-ibm replied:
I've removed it for the time being. It does cause the `integration` tests to fail at the end, but it's pretty obvious it fails because of a umount. Still, should investigate further

stevvooe commented on script/dind/dind:
```diff
@@ -0,0 +1,33 @@
+
+# Usage: dind CMD [ARG...]
+
+# apparmor sucks and Docker needs to know that it's in a container (c) @tianon
```
This is confusing to have all these mentions of docker but actually this is containerd in docker. Is this script still necessary? You might be able to get it running with just privileged and a few bind mounts.

stevvooe commented on script/setup/install-protobuf:
```diff
@@ -0,0 +1,27 @@
+GOARCH=$(go env GOARCH)
+PROTOBUF_DIR=$(mktemp -d)
+
+if [ $GOARCH == "amd64" ]; then
```
Looks like arm and windows have binary releases, as well.

stevvooe commented on script/setup/install-runc:
```diff
@@ -0,0 +1,13 @@
+go get -u github.com/opencontainers/runc
+cd $GOPATH/src/github.com/opencontainers/runc
+git checkout $RUNC_COMMIT
+make runc install
```
Does this need to be built with seccomp and apparmor enabled?

tophj-ibm commented on script/setup/install-protobuf:
```diff
@@ -0,0 +1,40 @@
+
+amd64|386)
+	if [ $GOOS == "windows" ]; then
+		wget -O $PROTOBUF_DIR/protobuf "https://github.com/google/protobuf/releases/download/v$PROTOBUF_VERSION/protoc-$PROTOBUF_VERSION-win32.zip"
```
I'm not 100% sure on this one, does 64-bit windows run win32 binaries?

stevvooe commented on script/setup/install-runc:
```diff
@@ -0,0 +1,13 @@
+#
+set -eu -o pipefail
+
+RUNC_COMMIT=$(grep opencontainers/runc ${GOPATH}/src/github.com/containerd/containerd/vendor.conf | cut -d " " -f 2)
```
Does this match what is in RUNC.md?

AkihiroSuda replied:
Generally no. I think we should replace RUNC.md with TOML/YAML that is readable for both machine and human. (Can be done in a separate PR)

estesp replied:
Well, it does match as long as PR changes are changing both locations. So far that is the case, but if we are worried about manual sync, then @AkihiroSuda might have a good idea to make a `runc` commit hash definition parseable. Of course, I assume we hope this goes away at some point as stable released `runc` will be viable for (hopefully) all releases of containerd.

crosbymichael commented:
LGTM
